### DISCUSSION TITLE
Patching CockroachDB cluster causing Hydra downtime when readiness probes fail

### DISCUSSION THREAD

**Discussion by dbhobbs** [ory/hydra#3057] [Q&A] [answered]
We are running a multi-region dedicated cluster of CockroachDB and have been running into downtime triggered by database patches. According to the [upgrade policy](https://www.cockroachlabs.com/docs/cockroachcloud/upgrade-policy#patch-version-upgrades) these patch updates may happen at any time and we have no control over them. However, this shouldn't cause any downtime as the database servers are removed from the cluster and patched one at a time and then added back to the cluster. The issue we're seeing is with our Hydra readiness probes failing. When a patch occurs the `/health/ready` check begins to timeout and eventually fail, which causes the Hydra pod to stop receiving traffic. This eventually affects _all_ Hydra pods and though they are still "alive", they aren't "ready" so we enter a period of downtime as no pods remain that can serve traffic. The strange thing is that all other operations with Hydra continue to work just fine, up until the point the pod stops receiving traffic because it isn't "ready".

Our `readinessProbe` is configured like [this](https://github.com/ory/k8s/blob/master/helm/charts/hydra/values.yaml#L279-L282):
```
readinessProbe:
    httpGet:
        path: /health/ready
        initialDelaySeconds: 30
        periodSeconds: 10
        failureThreshold: 5
```

I thought a discussion might be best before opening an issue as maybe we just have something configured incorrectly.

**Comment by aeneasr** [comment, marked as answer]
CRDB is notorious for issues when they do graceful rollovers. I have reported several issues to CRDB staff without resolve. The problem extends also to changefeeds and is not on Ory‘s side. One thing you can try is to tune the connection settings in your DSN (setting eg max age might help) or open a ticket with CRDB support

**Reply by dbhobbs** [reply]
Is it possible that this is related (and maybe even the fix)? #3095

**Reply by dbhobbs** [reply]
I've created a small repo that demonstrates what I believe we're seeing in our production Kubernetes cluster.
https://github.com/dbhobbs/crdb-hydra-test

**Reply by dbhobbs** [reply]
If my test repo above isn't completely off (which it very well may be :confused:).... I tried setting `max_conn_lifetime` to 10s but the `/health/ready` still _never_ reports healthy if the CRDB service that the hydra service is connected to goes down. I would have expected the connection to recycle and reconnect to one of the live nodes (as there are 4 still alive behind the load balancer) but I just get timeouts and errors until I bring that CRDB node back online.

**Reply by aeneasr** [reply]
@dbhobbs thank you for this reproduction repository, that is really great! Unfortunately, I think #3095 is not related to this as it touches the migration process, and not the HTTP server.

> I would have expected the connection to recycle and reconnect to one of the live nodes (as there are 4 still alive behind the load balancer) but I just get timeouts and errors until I bring that CRDB node back online.

Do you have the correct node configured in your DSN string? Or is it maybe always pointing to the same dead node? To my understanding, `max_conn_lifetime` should help here and establish a new connection. It's very strange that it apparently isn't.

**Reply by dbhobbs** [reply]
@aeneasr I was hoping, because `Ping` is called in the ready check, that it would reconnect the bad connection in the pool. But applying that patch results in the same behavior in my test repo.

Hydra's `dsn` is pointed to the HAProxy Load Balancer: `dsn: cockroach://hydra@dblb:5432/hydra?sslmode=disable&max_conns=20&max_idle_conns=4` (I've also tested with `max_conn_lifetime=10s`, as noted above). `dblb:5432` is the name and port of the HAProxy service.
https://github.com/dbhobbs/crdb-hydra-test/blob/main/configs/hydra/.hydra.yaml#L1

This behavior is confusing to me as the connection pool surely must know that the connection is dead and should recycle it and start a new connection through the LB, which _should_ reach a live db node.

**Reply by aeneasr** [reply]
In that case I think it's more likely that this is an issue with HAProxy, we use CRDB cloud with multiple nodes and while we had issues with CHANGEFEED connections in general there's no issues when one node goes down

**Reply by dbhobbs** [reply]
Turns out I was using the default values for HAProxy, which kept services in the pool much longer than I though. Adjusting some connection timeouts fixed some of the bad connection behavior I was seeing. I also had a conversation with an Enterprise Architect at Cockroach Labs and their recommendation is setting the `max_conn_lifetime` to 5m or less. Thanks for being patient with me.  :)
